articles/performance-optimization-techniques-for-node-js/index.md

```diff
@@ -0,0 +1,376 @@
+
+Switch to an external session store such as [Redis](https://redis.io/documentation), [Nginx](https://docs.nginx.com/nginx/deployment-guides/load-balance-third-party/node-js/#configuring-basic-session-persistence) or [MongoDB](https://docs.mongodb.com/manual/reference/method/Session/). Whenever possible, the best option would be to store no session state on your server-side for better performance.
+
+### Script Tracing and Logging
+Logging helps to track your application activities and traffic. When an app is running, it is possible to get fatal errors (even though your app was running properly after production testing). You need to get this feedback on time, find out what code has the error(s), and fix them before your user notices something is faulty in your system.
+
+Commonly used methods for logging in Node.js are `console.log()` which logs Standard outputs (stdout), and `console.error()`. They will log standard errors. However, there are more efficient, and scalable libraries/3rd party APIs that will assist you in logging Node.js scripts.
+
+They include:
+
+#### 1. Winston
+[Winston](https://github.com/winstonjs/winston) is a multi-transport async logger, which is simple, universal, and extremely versatile. It makes logging more flexible and extensible. It is the most popular logger based on NPM stats. Winston has different transports with different default levels that indicate message priority. [These levels include](https://github.com/winstonjs/winston#logging-levels)
+
+   - error
+   - warn
+   - info
+   - http
+   - verbose
+   - debug
+   - silly
+
+[**Example of a simple/custom Winston logger**](https://github.com/winstonjs/winston#creating-your-own-logger)
+
+```js
+const winston = require("winston");
+const logger = winston.createLogger({
+  transports: [new winston.transports.Console()],
+});
+logger.info("Information message");
+logger.warn("Warning message");
+logger.error("Error message");
+```
+
+[**Winston logger with file transport**](https://github.com/winstonjs/winston#multiple-transports-of-the-same-type)
+
+<iframe height="400px" width="100%" src="https://repl.it/@kimkimani/winston-logger-with-file-transport?lite=true" scrolling="no" frameborder="no" allowtransparency="true" allowfullscreen="true" sandbox="allow-forms allow-pointer-lock allow-popups allow-same-origin allow-scripts allow-modals"></iframe>
+
+#### 2. Morgan
+[Morgan](https://github.com/expressjs/morgan) is an HTTP request logger middleware for Node.js applications. Morgan gives insights on how your app is being used and alerts you on potential errors and issues that could be threats to your application. Morgan is considered the most reliable HTTP logger by Node.js developers. Morgan is designed to log errors the way servers like [Apache](https://httpd.apache.org/docs/1.3/logs.html) and [Nginx](https://docs.nginx.com/nginx/admin-guide/monitoring/logging/) carry out to the access-log or error-log.
```

Comment:
Usually, "insight" is used singularly. You use the word "insights" a couple of times in this article.